File: src/main/java/io/xol/chunkstories/api/entity/traits/serializable/TraitLocation.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.api.entity.traits.serializable;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.concurrent.locks.ReentrantLock;

import org.joml.Vector3dc;

import io.xol.chunkstories.api.Location;
import io.xol.chunkstories.api.entity.Entity;
import io.xol.chunkstories.api.events.entity.EntityTeleportEvent;
import io.xol.chunkstories.api.world.World;
import io.xol.chunkstories.api.world.WorldMaster;
import io.xol.chunkstories.api.world.chunk.Chunk;
import io.xol.chunkstories.api.world.serialization.StreamSource;
import io.xol.chunkstories.api.world.serialization.StreamTarget;

import javax.annotation.Nullable;

/** Holds the information about an entity whereabouts and a flag to mark it as
 * unspawned */
public class TraitLocation extends TraitSerializable {

	public TraitLocation(Entity entity, Location location) {
		super(entity);

		this.pos = location;
		this.world = location.getWorld();
	}

	private World world;
	private final Location pos;

	private final ReentrantLock lock = new ReentrantLock();

	@Nullable
	private Chunk chunk;

	private boolean spawned = false;
	private boolean removed = false;

	public void set(Location location) {
		if (location.world != this.world)
			throw new RuntimeException("Entities can't teleport between worlds directly.");

		set((Vector3dc) location);
	}

	public void set(Vector3dc position) {
		set(position.x(), position.y(), position.z());
	}

	public void set(double x, double y, double z) {
		EntityTeleportEvent event = new EntityTeleportEvent(entity, new Location(world, x, y, z));
		entity.world.getGameContext().getPluginManager().fireEvent(event);

		try {
			lock.lock();
			this.pos.x = (x);
			this.pos.y = (y);
			this.pos.z = (z);

			sanitize();
		} finally {
			lock.unlock();
		}

		// Push updates to everyone subscribed to this
		// In client mode it means that the controlled entity has the server subscribed
		// so it will update it's status to the server

		// In server/master mode any drastic location change is told to everyone, as
		// setLocation() is not called when the server receives updates
		// from the controller but only when an external event changes the location.
		this.pushComponentEveryone();
	}

	public void move(double mx, double my, double mz) {
		try {
			lock.lock();
			pos.x = (pos.x() + mx);
			pos.y = (pos.y() + my);
			pos.z = (pos.z() + mz);

			sanitize();
		} finally {
			lock.unlock();
		}

		this.pushComponentEveryone();
	}

	public void move(Vector3dc delta) {
		move(delta.x(), delta.y(), delta.z());
	}

	/** Copies the location and returns it. The actual location is never mutated
	 * outside of set(). */
	public Location get() {
		Location pos = this.pos;
		return new Location(pos.getWorld(), pos);
	}

	@Nullable
	public Chunk getChunk() {
		return chunk;
	}

	@Override
	public void push(StreamTarget to, DataOutputStream dos) throws IOException {
		dos.writeDouble(pos.x());
		dos.writeDouble(pos.y());
		dos.writeDouble(pos.z());
	}

	@Override
	public void pull(StreamSource from, DataInputStream dis) throws IOException {
		// pos = new Location(world, 0, 0, 0);

		double x = dis.readDouble();
		double y = dis.readDouble();
		double z = dis.readDouble();

		try {
			lock.lock();
			this.pos.x = x;
			this.pos.y = y;
			this.pos.z = z;

			sanitize();
		} finally {
			lock.unlock();
		}

		// Position updates received by the server should be told to everyone but the
		// controller
		if (world instanceof WorldMaster)
			pushComponentEveryoneButController();
	}

	/** Prevents entities from going outside the world area and updates the
	 * parentHolder reference */
	protected final boolean sanitize() {
		double worldSize = world.getWorldSize();

		pos.x = pos.x() % worldSize;
		pos.z = pos.z() % worldSize;

		// Loop arround the world
		if (pos.x() < 0)
			pos.x = pos.x() + worldSize;
		else if (pos.x() > worldSize)
			pos.x = pos.x() % worldSize;

		if (pos.z() < 0)
			pos.z = pos.z() + worldSize;
		else if (pos.z() > worldSize)
			pos.z = pos.z() % worldSize;

		if (pos.y < 0)
			pos.y = 0;

		if (pos.y > world.getMaxHeight())
			pos.y = world.getMaxHeight();

		// Get local chunk co-ordinate
		int chunkX = ((int) pos.x()) >> 5;
		int chunkY = ((int) pos.y()) >> 5;
		int chunkZ = ((int) pos.z()) >> 5;

		// Don't touch updates once the entity was removed
		if (removed /* !entity.exists() */)
			return false;

		// Entities not in the world should never be added to it
		if (!spawned)
			return false;

		if (chunk != null && chunk.getChunkX() == chunkX && chunk.getChunkY() == chunkY && chunk.getChunkZ() == chunkZ) {
			return false; // Nothing to do !
		} else {
			if (chunk != null)
				chunk.removeEntity(entity);

			chunk = world.getChunk(chunkX, chunkY, chunkZ);
			// When the region is loaded, add this entity to it.
			if (chunk != null)// && regionWithin.isDiskDataLoaded())
				chunk.addEntity(entity);

			return true;
		}
	}

	public void onRemoval() {
		try {
			lock.lock();

			if (chunk != null)
				chunk.removeEntity(entity);

			removed = true;
		} finally {
			lock.unlock();
		}

		this.pushComponentEveryone();

		// Tell anyone still subscribed to this entity to sod off
		entity.subscribers.all().forEach(subscriber -> {
			subscriber.unsubscribe(entity);
		});
	}

	public boolean wasRemoved() {
		return removed;
	}

	public void onSpawn() {
		try {
			lock.lock();
			spawned = true;

			sanitize();
		} finally {
			lock.unlock();
		}
	}

	public boolean hasSpawned() {
		return spawned;
	}
}

File: src/main/java/xyz/chunkstories/api/entity/traits/serializable/TraitLocation.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.entity.traits.serializable

import org.joml.Vector3dc
import xyz.chunkstories.api.Location
import xyz.chunkstories.api.content.json.Json
import xyz.chunkstories.api.content.json.asArray
import xyz.chunkstories.api.content.json.asDouble
import xyz.chunkstories.api.entity.Entity
import xyz.chunkstories.api.entity.Subscriber
import xyz.chunkstories.api.entity.traits.Trait
import xyz.chunkstories.api.player.Player
import xyz.chunkstories.api.world.WorldMaster
import java.io.DataInputStream
import java.io.DataOutputStream

/** Holds the information about an entity whereabouts and a flag to mark it as unspawned  */
class TraitLocation(entity: Entity, private val actualLocation: Location) : Trait(entity), TraitSerializable, TraitNetworked<TraitLocation.LocationUpdate> {
    override val traitName = "location"

    fun set(location: Location) {
        if (location.world !== entity.world)
            throw RuntimeException("Entities can't teleport between worlds directly.")
        set(location as Vector3dc)
    }

    fun set(position: Vector3dc) {
        set(position.x(), position.y(), position.z())
    }

    operator fun set(x: Double, y: Double, z: Double) {
        this.actualLocation.x = x
        this.actualLocation.y = y
        this.actualLocation.z = z

        // Push updates to everyone subscribed to this
        // In client mode it means that the controlled entity has the server subscribed
        // so it will update it's status to the server

        // In server/master mode any drastic location change is told to everyone, as
        // setLocation() is not called when the server receives updates
        // from the controller but only when an external event changes the location.
        sendMessageAllSubscribers(LocationUpdate(actualLocation.x, actualLocation.y, actualLocation.z))
    }

    fun move(dx: Double, dy: Double, dz: Double) {
        if (dx.isNaN() || dy.isNaN() || dz.isNaN()) {
            System.err.println("move by nan $dx $dy $dz")
            Thread.dumpStack()
            return
        }

        actualLocation.x = actualLocation.x() + dx
        actualLocation.y = actualLocation.y() + dy
        actualLocation.z = actualLocation.z() + dz

        sendMessageAllSubscribers(LocationUpdate(actualLocation.x, actualLocation.y, actualLocation.z))
    }

    fun move(delta: Vector3dc) {
        move(delta.x(), delta.y(), delta.z())
    }

    /** Copies the location and returns it. The actual location is never mutated
     * outside of set().  */
    fun get(): Location {
        val pos = this.actualLocation
        return Location(pos.world, pos)
    }

    data class LocationUpdate(val x: Double, val y: Double, val z: Double) : TraitMessage() {
        override fun write(dos: DataOutputStream) {
            dos.writeDouble(x)
            dos.writeDouble(y)
            dos.writeDouble(z)
        }
    }

    override fun readMessage(dis: DataInputStream): LocationUpdate =
            LocationUpdate(dis.readDouble(), dis.readDouble(), dis.readDouble())

    override fun processMessage(message: LocationUpdate, player: Player?) {
        if (entity.world is WorldMaster && player != entity.controller) {
            throw Exception("Security violation: Someone tried to update an entity they don't control !")
        }

        this.actualLocation.x = message.x
        this.actualLocation.y = message.y
        this.actualLocation.z = message.z

        // Position updates received by the server should be told to everyone but the controller
        if (entity.world is WorldMaster)
            sendMessageAllSubscribersButController(message)
    }

    override fun tick() {
        if (get().x.isNaN() || get().y.isNaN() || get().z.isNaN()) {
            entity.world.logger.warn("Entity $entity had invalid location: ${get()}, resetting to world spawn")
            set(entity.world.properties.spawn)
        }
    }

    override fun whenSubscriberRegisters(subscriber: Subscriber) {
        sendMessage(subscriber, LocationUpdate(actualLocation.x, actualLocation.y, actualLocation.z))
    }

    override fun deserialize(json: Json) {
        val arr = json.asArray ?: return
        actualLocation.set(arr[0].asDouble!!, arr[1].asDouble!!, arr[2].asDouble!!)
    }

    override fun serialize(): Json {
        return Json.Array(listOf(Json.Value.Number(actualLocation.x), Json.Value.Number(actualLocation.y), Json.Value.Number(actualLocation.z)))
    }
}

fun Entity.move(delta: Vector3dc) = this.traits[TraitLocation::class]?.move(delta)
fun Entity.move(dx: Double, dy: Double, dz: Double) = this.traits[TraitLocation::class]?.move(dx, dy, dz)